### ISSUE TITLE
Recommended systemd configuration for v2.2?

### ISSUE THREAD

**Issue by canadaduane** [rvaiya/keyd#104]
I'm exploring having `keyd` and its companion `keyd-application-mapper` run as a "system-wide" systemd unit and a "user" systemd unit, respectively. (Note: on my Pop!_OS / ubuntu system, I can't run `keyd-application-mapper` via `.xinitrc`, as it does not trigger during a gdm session).

I'm not an expert in systemd configs, but I _think_ that the new socket capability of keyd in v2.2 means that the ideal systemd configuration should declare a socket (see below).

As far as I understand it, systemd's methodology is to create a socket _first_, even before the services that read or write to the socket have been called.

Unfortunately, keyd currently checks to see if the socket already exists, and if it does, it throws an error:

```
Jan 16 10:07:37 pop-os-fw1 keyd[21186]: src/keyd.c:916: ERROR: Another instance of keyd appears to be runnin>
Jan 16 10:07:37 pop-os-fw1 keyd[21186]:         to force keyd to run, manually remove /var/run/keyd.socket a>
Jan 16 10:07:37 pop-os-fw1 sudo[21183]: pam_unix(sudo:session): session closed for user root
Jan 16 10:07:37 pop-os-fw1 systemd[1]: keyd.service: Main process exited, code=exited, status=255/EXCEPTION
```

Should `keyd` use another mechanism to determine if another instance is running?

Here are my systemd config files:

`/etc/systemd/system/keyd.socket`:
```ini
[Unit]
Description=Key Remapping Daemon Socket
PartOf=keyd.service

[Socket]
ListenStream=/var/run/keyd.socket
SocketMode=0660

[Install]
WantedBy=sockets.target
```

The socket accompanies the daemon service unit:

`/etc/systemd/system/multi-user.target.wants/keyd.service`:
```ini
[Unit]
Description=Key Remapping Daemon
Requires=keyd.socket

[Service]
Type=exec
ExecStart=/usr/bin/keyd 

[Install]
WantedBy=multi-user.target
```

Finally, the "user" level service:

`~/.config/systemd/user/keyd-application-mapper.service`:
```ini
[Unit]
Description=keyd active application monitor

[Service]
Type=exec
ExecStart=/usr/bin/keyd-application-mapper

[Install]
WantedBy=multi-user.target
```

Note for future readers: the above setup assumes a `keyd` group exists on the system (`sudo groupadd keyd`) and that the user has the `keyd` group added (`sudo usermod -a -G keyd $(whoami)`). It also assumes `keyd` and `keyd-application-mapper` are installed at `/usr/bin/keyd`, and that conf files are in `/etc/keyd/*.conf` for system-level keyd daemon, and `~/.config/keyd/app.conf` for user-level keyd-application-mapper.

**Comment by canadaduane** [comment]
Note: I've commented out lines 915-917 and it appears to be working for me:
```
	chgid();
	// if (!access(SOCKET, F_OK))
	// 	die("ERROR: Another instance of keyd appears to be running.\n"
	// 	     "\tto force keyd to run, manually remove %s and try again.", SOCKET);
```

**Comment by rvaiya** [comment]
> I'm exploring having keyd and its companion keyd-application-mapper run as a "system-wide" systemd unit and a "user" systemd unit, respectively. (Note: on my Pop!_OS / ubuntu system, I can't run keyd-application-mapper via .xinitrc, as it does not trigger during a gdm session).

> As far as I understand it, systemd's methodology is to create a socket first, even before the services that read or write to the socket have been called.

systemd does have such a facility, but there isn't really a compelling reason to use it in this case. The ostensible advantage of using systemd to manage sockets is that it allows interdependent processes to be launched concurrently by allowing them to block on pre-initialized sockets until their dependencies are created. In the case of keyd this just adds a bunch of complexity for no real gain and makes systemd a hard dependency (which is undesirable for a number of reasons). There are many (most?) programs which don't use this feature.

> Should keyd use another mechanism to determine if another instance is running?

This particular issue can be fixed either by using a PID file or by adding a simple flock check (the old behaviour). I don't remember why I didn't do this. I think my rationale was that if the socket was still lying around then keyd had probably been killed irregularly (it should clean up on SIGTERM and SIGINT) and it would be desirable for the user to be explicitly aware of this. I may also have just been lazy :P.

```
// if (!access(SOCKET, F_OK))
	// 	die("ERROR: Another instance of keyd appears to be running.\n"
	// 	     "\tto force keyd to run, manually remove %s and try again.", SOCKET);
```

This allows multiple instances to be run at once.

```
~/.config/systemd/user/keyd-application-mapper.service:

[Unit]
Description=keyd active application monitor

[Service]
Type=exec
ExecStart=/usr/bin/keyd-application-mapper

[Install]
WantedBy=multi-user.target
````

I'm not sure when gdm and company initialize systemd user units but it is not uncommon for users of other window managers to log in before the display server has been initialized (i.e directly into the TTY). In general it is probably not a good idea to initialize something which depends on the display server in a systemd unit. 

In X `~/.xinitrc` can be used for initialization, but (as usual) there doesn't seem to be a consensus in in the wayland world. Apparently gnome ships with something called the 'tweak tool' which can be used to add run scripts upon start up. In the case of gnome it may make sense to do initialization in the installed extension, but this would probably mean enabling verbose mode by default and constantly growing the log (which is probably not a big deal in practice).

**Comment by canadaduane** [comment]
> In X ~/.xinitrc can be used for initialization
I don't think this is universally true. At least, on my system, it is not the case that `.xinitrc` can be used to run programs at start of the X session.

**Comment by rvaiya** [comment]
You're right, `xinitrc` only applies if you launch via `startx`. Apparently `~/.xsession` is the standard for display managers (I don't use one).

**Comment by canadaduane** [comment]
So the thing that I don't understand then is: how can we guarantee that `keyd` has started before whatever we run in `.xprofile` or `.xinitrc`, if there is no systemd involved making sure services don't start before others are available?

Previously, when I ran this repository's suggested https://github.com/rvaiya/keyd/blob/master/keyd.service it would occasionally fail to start. I didn't investigate why at the time, because I didn't have enough systemd knowledge to debug it. Maybe I need to now.

**Comment by canadaduane** [comment]
> systemd does have such a facility, but there isn't really a compelling reason to use it in this case. The ostensible advantage of using systemd to manage sockets is that it allows interdependent processes to be launched concurrently by allowing them to block on pre-initialized sockets until their dependencies are created.

But this is precisely what we're doing now: we have two services (`keyd` and `keyd-application-mapper`) that must either load in series, slightly slowing down initialization OR we can load them in parallel only if the socket exists before both services start. It seems like a nice-to-have feature here that systemd is offering, if keyd will support it.

> In the case of keyd this just adds a bunch of complexity for no real gain and makes systemd a hard dependency (which is undesirable for a number of reasons). There are many (most?) programs which don't use this feature.

I don't think anything I've done makes systemd a hard dependency, does it?

**Comment by rvaiya** [comment]
> So the thing that I don't understand then is: how can we guarantee that keyd has started before whatever we run in .xprofile or .xinitrc, if there is no systemd involved making sure dependencies of services start beforehand?

If the user is using the supplied systemd service then keyd should already be running at start up. In general it is assumed that the user will want access to their keyboard configuration as early as possible, which means they are unlikely to manually start the service. The script as it is currently written just does an initial check for the existence of the socket as a courtesy, but this can be removed without issue since invocations of `keyd -e` will fail silently until the daemon is running.

> Previously, when I ran this repository's suggested https://github.com/rvaiya/keyd/blob/master/keyd.service it would occasionally fail to start. I didn't investigate why at the time, because I didn't have enough systemd knowledge to debug it. Maybe I need to now.

This should be considered a bug that needs to be fixed. Next time it fails to start you can obtain logs with `journalctl -fu keyd` and create an issue.
> But this is precisely what we're doing now: we have two services (keyd and keyd-application-mapper) that must either load in series (slightly slowing down initialization)

The script already issues a warning if the system service isn't running and the user should be aware when/if the system process is running since it manages the keymap. The script shouldn't be managed by systemd anyway for the reasons mentioned in the earlier post, but even if you did make it a user level service it wouldn't be able to automatically start the system level one. This has nothing to do with socket creation.

> I don't think anything I've done makes systemd a hard dependency, does it? Perhaps I'm misunderstanding something.

I believe systemd has facilities for passing the created socket (which is defined in the unit file) into the running program which requires the program to be patched (e.g sd_listen_fds(3)). In any event, even if systemd is just used to create the socket file this still creates an external dependency. All initialization should be done by the daemon itself with minimal knowledge of the init system.

Since systemd intergration seems to be a perennial issue, it is worth noting that systemd is a relatively recent and controversial development in the linux ecosystem. While it seems to have mostly won the init wars (in linux land) it still has its share of detractors and there are many people who see it as a violation of the unix philosophy and refuse to use it.

https://nosystemd.org/

I should also note that I am not expressing a personal opinion here :P.

**Comment by canadaduane** [comment]
Ok, I've reverted my setup to your recommended `keyd.service` with a line in my `.xprofile` that starts `keyd-application-mapper`. This works fine. I think we can close this issue, and I'll report back if I have any further problems with `keyd` failing to load at boot.
